cpp/test/sg/fnv_hash_test.cpp

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuml/fil/fnv_hash.h>
+#include <gtest/gtest.h>
+#include <raft/error.hpp>
+
+typedef std::tuple<std::vector<char>, unsigned long long, unsigned long>
+  fnv_vec_t;
+
+class FNVHashTest : public testing::TestWithParam<fnv_vec_t> {
+ protected:
+  void SetUp() override { std::tie(input, correct64, correct32) = GetParam(); }
+
+  void check() {
+    unsigned long long real64 =
+      fowler_noll_vo_fingerprint64(input.begin(), input.end());
+    ASSERT(real64 == correct64, "Wrong hash computed");
+    unsigned long real32 =
+      fowler_noll_vo_fingerprint64_32(input.begin(), input.end());
+    ASSERT(real32 == correct32, "Wrong xor-folded hash computed");
+  }
+
+  // parameters
+  std::vector<char> input;
+  unsigned long long correct64;
+  unsigned long correct32;
+};
+
+std::vector<fnv_vec_t> fnv_vecs = {
+  {{}, 14695981039346656037ull, 0xcbf29ce4 ^ 0x84222325},  // test #0
+  {{0}, 0xaf63bd4c8601b7df, 0xaf63bd4c ^ 0x8601b7df},
+  {{1}, 0xaf63bd4c8601b7de, 0xaf63bd4c ^ 0x8601b7de},
```

Comment:
Why is the 32-bit output always provided as a bit-wise xor of 2 different constants? A single constant would be better.

Reply:
because it's intended to be the xor of two halves of the 64-bit output. This is obvious if you look at the numbers in this format, else the test vector will need to be verified manually. Shall I leave a comment about this in the code?